spring-kafka/src/test/java/org/springframework/kafka/core/DefaultShareConsumerFactoryTests.java

```diff
@@ -197,4 +185,112 @@ void integrationTestDefaultShareConsumerFactory(EmbeddedKafkaBroker broker) thro
+		final String topic = "embedded-share-distribution-test";
+		final String groupId = "distributionTestGroup";
+		int recordCount = 8;
+		List<String> consumerIds = List.of("client-dist-1", "client-dist-2");
+		Map<String, Set<String>> consumerRecords = runSharedConsumerTest(topic, groupId, consumerIds,
+				recordCount, broker);
+
+		// Assert all records were received (no loss)
+		Set<String> allReceived = new java.util.HashSet<>();
+		consumerRecords.values().forEach(allReceived::addAll);
+		for (int i = 0; i < recordCount; i++) {
+			assertThat(allReceived)
+				.as("Should have received value " + topic + "-value-" + i)
+				.contains(topic + "-value-" + i);
+		}
+	}
+
+	/**
+	 * Runs multiple Kafka consumers in parallel using ExecutorService, collects all records received,
+	 * and returns a map of consumerId to the set of record values received by that consumer.
+	 */
+	private static Map<String, Set<String>> runSharedConsumerTest(String topic, String groupId,
+			List<String> consumerIds, int recordCount, EmbeddedKafkaBroker broker) throws Exception {
+		var bootstrapServers = broker.getBrokersAsString();
+
+		var producerProps = new java.util.Properties();
+		producerProps.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
+		producerProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+		producerProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+		try (var producer = new KafkaProducer<String, String>(producerProps)) {
+			for (int i = 0; i < recordCount; i++) {
+				producer.send(new ProducerRecord<>(topic, "key" + i, topic + "-value-" + i)).get();
+			}
+			producer.flush();
+		}
+
+		setShareAutoOffsetResetEarliest(bootstrapServers, groupId);
+
+		Map<String, Set<String>> consumerRecords = new java.util.concurrent.ConcurrentHashMap<>();
+		consumerIds.forEach(id -> consumerRecords.put(id,
+				java.util.Collections.synchronizedSet(new java.util.HashSet<>())));
+		var latch = new java.util.concurrent.CountDownLatch(recordCount);
+		var running = new java.util.concurrent.atomic.AtomicBoolean(true);
+		ExecutorService executor = Executors.newCachedThreadPool();
+		List<Future<?>> futures = new java.util.ArrayList<>();
+
+		// Consumer task: poll, acknowledge, and count down latch for new records
+		for (int i = 0; i < consumerIds.size(); i++) {
+			final int idx = i;
+			Future<?> future = executor.submit(() -> {
+				DefaultShareConsumerFactory<String, String> shareConsumerFactory = new DefaultShareConsumerFactory<>(
+						Map.of(
+							"bootstrap.servers", bootstrapServers,
+							"key.deserializer", org.apache.kafka.common.serialization.StringDeserializer.class,
+							"value.deserializer", org.apache.kafka.common.serialization.StringDeserializer.class
+						));
+				try (var consumer = shareConsumerFactory
+						.createShareConsumer(groupId, consumerIds.get(idx))) {
```

Comment:
Any reason to create consumers in those threads?
Why cannot we do that outside for the ` executor.submit()`?

Reply:
OK. But that is not what I see in the `KafkaMessageListenerContainer`.
We create a `ListenerConsumer` from the `doStart()` and immediately there we call `consumerFactory.createConsumer()`.
However the `poll()` we indeed call from the other thread.
Therefore my confusion here why would we create this shared consumer instance exactly in a thread where we poll?

Reply:
In the framework code you mentioned in `KafkaMessageListenerContainer`, although we hand off the created consumer to another thread, in the `doStart()` method, we ensure that the created consumer instance is exactly what is handed off to the polling thread and therefore it is thread safe. However, in our test scenario here, unless we do the same thing, (i.e., create a dedicated consumer instance and then pass that exactly to the thread), we may run into thread safety issues. IMO, it might be okay to to do what we do right now in the test. But if you see any issues with that, I will try to refactor the test.